File: compat/suites/python-sdk/lib/scenario/executor.py
"""
lib/scenario/executor.py — making the calls a scenario describes.

The SDK is used exactly as production code would use it: ``boto3.client(svc)``
and ``getattr(client, xform_name(op))(**params)`` is boto3's ordinary public
API, which is the whole reason ``python-sdk`` gets an interpreter rather than
generated source (plan §3.2). There is no Overcast-specific code path here; the
endpoint override lives in ``lib/clients.py`` and is the only deviation from a
production client.

Everything that can fail funnels through :class:`Executor`, so that a call, an
unresolvable ``$ref`` and a missing export path all report with the same six
fields as an assertion failure.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, replace
from typing import Any, Callable, Optional

from botocore import xform_name

from ..clients import make_client
# _is_unimplemented is the harness's own 501 detection, and the interpreter has
# to defer to it rather than re-implement it: what counts as "the emulator has
# not implemented this" must be one rule for the whole suite, or a generated
# probe and a hand-written test would disagree about the same response.
from ..harness import TestContext, _is_unimplemented
from .expressions import evaluate, resolve_path
from .failures import MISSING, ScenarioError, ScenarioFailure, failure_message
from .loader import ScenarioGroup

# Where the group's context bag lives on the TestContext. The harness gives
# every group its own TestContext, and setup, tests and teardown of one group
# all receive that same object — which is exactly the lifetime the IR gives the
# context bag.
_BAG_KEY = "__scenario_context__"


# The whole of compat/model/README.md § Naming's python-sdk row: botocore's
# service name is the scenario's endpoint prefix except for four services,
# where botocore keeps a shorter historical directory name. It is a table
# because nothing about "elasticloadbalancing" implies "elb" — botocore's
# service names come from its own data directory, not from the endpoint, and
# for these four the two disagree. ``boto3.client("elasticloadbalancing")``
# raises UnknownServiceError, so a generated group for elastic-load-balancing
# would report every test as a failure of the service rather than of the
# derivation.
#
# The plan (§7.3) asked for this to land with the first scenario that names one
# of the four rather than up front, which is ``elastic-load-balancing``; the
# other three are here because they are one documented set and a table with one
# entry invites the next author to add theirs somewhere else.
_BOTOCORE_SERVICE_OVERRIDES = {
    "elasticloadbalancing": "elb",
    "monitoring": "cloudwatch",
    "email": "ses",
    "states": "stepfunctions",
}


def botocore_service(endpoint_prefix: str) -> str:
    """The botocore service name for a scenario's endpoint prefix."""
    return _BOTOCORE_SERVICE_OVERRIDES.get(endpoint_prefix, endpoint_prefix)


@dataclass(frozen=True)
class StepRef:
    """Fields 1 and 6 of a failure message: which test failed, and where in
    the scenario file the step that failed is written.

    ``test`` is a test name for a test, and ``setup``/``teardown`` for the
    group-level phases, which have no test of their own."""

    group: str
    test: str
    file: str
    step: str

    def child(self, step: str) -> "StepRef":
        """A nested step — the clause inside an ``eventually``."""
        return replace(self, step=f"{self.step}.{step}")


class Executor:
    """Executes calls for one group, against one context bag."""

    def __init__(self, spec: ScenarioGroup, ctx: TestContext, clients: "ClientCache") -> None:
        self.spec = spec
        self.ctx = ctx
        self._clients = clients
        bag = ctx.get(_BAG_KEY)
        if bag is None:
            bag = {}
            ctx[_BAG_KEY] = bag
        self.context: dict[str, Any] = bag

    # ── expressions ─────────────────────────────────────────────────────────

    def evaluate(self, value: Any) -> Any:
        return evaluate(value, context=self.context, run_id=self.ctx.run_id,
                        group=self.spec.name)

    def params_for(self, call: dict, ref: StepRef, assertion: str) -> dict:
        """Evaluate a call's params, or fail naming the expression that could
        not be evaluated. The raw IR params are shown, because the evaluated
        ones are what does not exist."""
        try:
            return self.evaluate(call.get("params") or {})
        except ScenarioError as exc:
            raise ScenarioFailure(failure_message(
                group=ref.group, test=ref.test, op=call["op"],
                params=call.get("params") or {}, assertion=assertion,
                expected="every value expression resolves",
                actual=str(exc),
                scenario_file=ref.file, step=ref.step,
            )) from exc

    # ── calls ───────────────────────────────────────────────────────────────

    def invoke(self, call: dict, params: dict) -> dict:
        """The SDK call itself, with nothing wrapped around it.

        Params are passed as the IR gives them: the generator emits a value of
        each member's modeled kind (a string for a string member, a number for
        a numeric one), so botocore's own input validation is the only coercion
        in play — the same one a hand-written call gets. Verified against
        botocore's service models for the whole pilot corpus; if a future
        scenario ever did carry a string for a numeric member, botocore would
        reject it here rather than silently coerce, which is the right failure.
        """
        client = self._clients.get(self.ctx, botocore_service(self.spec.client["endpointPrefix"]))
        return getattr(client, xform_name(call["op"]))(**params)

    def perform(self, call: dict, ref: StepRef, assertion: str, *,
                apply_exports: bool = True) -> tuple[dict, dict]:
        """Evaluate params, make the call, apply the call's exports. Returns
        ``(params, response)``.

        ``apply_exports=False`` is for a clause whose exports are conditional
        on the clause holding — a ``readback``'s are applied only when its
        checks pass, so that a failed attempt inside an ``eventually`` cannot
        leave a half-written context bag behind.

        An error that is the emulator answering 501 is re-raised unchanged, so
        the harness's unimplemented detection still sees a botocore
        ``ClientError`` and records ``unimplemented`` rather than ``fail`` —
        which is the whole point of the probe groups."""
        params = self.params_for(call, ref, assertion)
        try:
            response = self.invoke(call, params)
        except Exception as exc:
            if _is_unimplemented(exc):
                raise
            raise ScenarioFailure(failure_message(
                group=ref.group, test=ref.test, op=call["op"], params=params,
                assertion=assertion, expected="the call succeeds",
                actual=f"{type(exc).__name__}: {exc}",
                scenario_file=ref.file, step=ref.step,
            )) from exc
        if apply_exports:
            self.apply_export(call, params, response, ref, assertion)
        return params, response

    def attempt(self, call: dict, ref: StepRef, assertion: str
                ) -> tuple[dict, Optional[dict], Optional[Exception]]:
        """Like :meth:`perform`, but for a clause that *expects* an error:
        returns ``(params, response, error)`` with exactly one of the last two
        set. Exports are not applied — a call that was supposed to fail has
        nothing to export, and one that succeeded is about to fail its clause.
        """
        params = self.params_for(call, ref, assertion)
        try:
            return params, self.invoke(call, params), None
        except Exception as exc:  # noqa: BLE001 - the clause decides what it means
            if _is_unimplemented(exc):
                raise
            return params, None, exc

    # ── exports ─────────────────────────────────────────────────────────────

    def apply_export(self, call: dict, params: dict, response: dict, ref: StepRef,
                     assertion: str) -> None:
        """Write a call's ``export`` paths into the context bag. A path the
        response does not carry is a failure of the step that carries it,
        naming the path."""
        for name, path in (call.get("export") or {}).items():
            value = resolve_path(response, path)
            if value is MISSING:
                raise ScenarioFailure(failure_message(
                    group=ref.group, test=ref.test, op=call["op"], params=params,
                    assertion=assertion, path=path,
                    expected=f'the response carries {path}, to export as "{name}"',
                    actual="<missing>",
                    scenario_file=ref.file, step=ref.step,
                ))
            self.context[name] = value

    # ── message helper ──────────────────────────────────────────────────────

    def fail(self, *, ref: StepRef, op: str, params: Any, assertion: str,
             expected: str, actual: str, path: Optional[str] = None) -> ScenarioFailure:
        return ScenarioFailure(failure_message(
            group=ref.group, test=ref.test, op=op, params=params,
            assertion=assertion, path=path, expected=expected, actual=actual,
            scenario_file=ref.file, step=ref.step,
        ))


class ClientCache:
    """One boto3 client per (endpoint, region, service), shared across the
    groups of a run.

    Building a client parses botocore's service model, which is slow enough to
    matter across 34 organizations tests, and a botocore client is safe to call
    from several threads once built. Construction is not, hence the lock."""

    def __init__(self, factory: Callable[[str, str, str], Any] = make_client) -> None:
        self._factory = factory
        self._lock = threading.Lock()
        self._cache: dict[tuple[str, str, str], Any] = {}

    def get(self, ctx: TestContext, service: str) -> Any:
        key = (ctx.endpoint, ctx.region, service)
        with self._lock:
            client = self._cache.get(key)
            if client is None:
                client = self._factory(ctx.endpoint, ctx.region, service)
                self._cache[key] = client
            return client


def error_names(exc: Exception) -> list[str]:
    """Every name an SDK error reports itself under.

    The SDKs disagree about whether they surface the modeled shape name or the
    wire code, so compat/model/README.md § Errors has an interpreter accept
    either against either, over a fixed list of surfaces. This reads the three
    boto3 actually has:

    * the exception class botocore minted, which is the modeled shape's name
      whenever ``Error.Code`` matched a modeled error's code;
    * ``Error.Code`` — one field, wherever the code came from. botocore
      resolves it from the body's ``__type`` (namespace already stripped), from
      ``x-amzn-errortype`` or the body's ``code``/``Code`` member for REST JSON,
      from the ``Code`` inside the XML error node for AWS Query and REST XML —
      ``<ErrorResponse><Error><Code>`` and the bare ``<Error><Code>`` alike,
      which is the nested half of the Errors table's body-code row and the only
      place a Query service ever states its code — and replaces it with the
      ``x-amzn-query-error`` code for a query-compatible service. This one
      field is why the nested position needs no separate read here: botocore
      has already put it where the top-level one goes. There is no
      ``Error.__type`` and no top-level ``__type`` to read: botocore never sets
      either, so looking for them would be reading a key only a test fixture
      could put there. ``Error.code`` is read because the Errors table names
      that spelling, but botocore does not use it;
    * the ``x-amzn-query-error`` header itself, which is on the response
      whether or not the parser preferred it.

    Each is read in every spelling ``_spellings`` lists. The shared fixtures
    under ``compat/model/testdata/errors`` pin the outcomes;
    ``tests/test_error_fixtures.py`` runs them against this."""
    names = [type(exc).__name__]
    response = getattr(exc, "response", None)
    if isinstance(response, dict):
        error = response.get("Error") or {}
        for key in ("Code", "code"):
            names.extend(_spellings(error.get(key)))
        headers = (response.get("ResponseMetadata") or {}).get("HTTPHeaders") or {}
        names.extend(_spellings(headers.get("x-amzn-query-error")))
    seen: list[str] = []
    for name in names:
        if name and name not in seen:
            seen.append(name)
    return seen


def _spellings(value: Any) -> list[str]:
    """One raw surface, in every spelling a clause may name it by: the value
    itself, what follows the last ``#`` of a Smithy id, and what precedes the
    first ``;`` of the header's ``<code>;<fault>`` form. Splitting only at
    those separators is what keeps matching an equality: no spelling of
    ``ResourceNotFoundException`` is ever ``NotFoundException``."""
    if not isinstance(value, str) or not value:
        return []
    out = [value]
    if "#" in value:
        out.append(value.rsplit("#", 1)[-1])
    if ";" in value:
        out.append(value.split(";", 1)[0])
    return out


def describe_error(exc: Optional[Exception]) -> str:
    """The fifth field's "actual" half for an error clause."""
    if exc is None:
        return "<no error>"
    return f"{'/'.join(error_names(exc))}: {exc}"


def error_matches(exc: Exception, error: dict) -> bool:
    accepted = {error.get("shape"), error.get("code")} - {None}
    return any(name in accepted for name in error_names(exc))


__all__ = [
    "ClientCache",
    "Executor",
    "StepRef",
    "describe_error",
    "error_matches",
    "error_names",
]

File: compat/suites/node-js-sdk/src/lib/scenario/executor.ts
/**
 * executor.ts — run a scenario group's setup, tests and teardown.
 *
 * The order compat/model/README.md § The scenario file fixes:
 *
 *   1. an empty context;
 *   2. every setup call — a failure reports every test as `skip` with
 *      `setup failed: <message>`, where `<message>` is the failing step's own
 *      six-field failure message, and teardown still runs;
 *   3. every test in order (the registry's `depends` gives the loader the
 *      order and the usual dependency skip);
 *   4. every teardown call, each individually wrapped: an error or an
 *      unresolvable `$ref` skips that call and continues with the next.
 *
 * Steps 2 and 4 are wired to harness.ts's own `setup`/`teardown` hooks, which
 * already implement the skip-the-group and always-run halves of the rule, so
 * this module supplies the calls and nothing else.
 */

import {
  ExpressionError,
  evaluateParams,
  resolvePath,
  toDocument,
} from "./expressions.ts";
import type { EvalContext } from "./expressions.ts";
import {
  acceptedCodes,
  errorMatches,
  evaluateChecks,
  evaluateListAbsent,
  evaluateListContains,
} from "./assertions.ts";
import type { Mismatch } from "./assertions.ts";
import { describeError, failureMessage, scenarioFailure } from "./failure.ts";
import type { FailureSite } from "./failure.ts";
import type { Assertion, Call, ScenarioTest } from "./ir.ts";
import type { Sender } from "./client.ts";

/** Everything a step needs that is not in the scenario file. */
export interface ExecEnv {
  send: Sender;
  /** runId, group name and the context bag. */
  ctx: EvalContext;
  /** Repo-relative scenario path, for field 6 of a failure message. */
  scenarioFile: string;
  /** Debug output; harness's ctx.log (stderr, never stdout). */
  log: (msg: string) => void;
  /** Injected so tests do not wait in real time. */
  sleep: (ms: number) => Promise<void>;
}

/** The default sleep: a real timer. */
export function realSleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

interface CallResult {
  /** The evaluated params, exactly as sent — field 3 of a failure message. */
  params: Record<string, unknown>;
  response: unknown;
}

/** The primary call's outcome, shared by every clause of the test. */
interface Primary {
  op: string;
  params: Record<string, unknown>;
  response: unknown;
  /** Set when the call threw — only legal for a test with an errorCode clause. */
  error?: unknown;
}

// ─── Setup and teardown ───────────────────────────────────────────────────

/**
 * Run every setup call in order.
 *
 * Throws the failure message as a bare string rather than an Error so that
 * harness.ts's `setup failed: ${String(err)}` reads exactly as the contract
 * spells it, instead of gaining an "Error: " prefix. Same reasoning, and same
 * shape, as registry.ts's scenarioBackendMissing().
 */
export async function runSetup(env: ExecEnv, calls: Call[]): Promise<void> {
  for (const [i, call] of calls.entries()) {
    const site: FailureSite = {
      group: env.ctx.group,
      test: "<setup>",
      scenarioFile: env.scenarioFile,
      step: `setup[${i}]`,
    };
    try {
      const result = await performCall(env, site, call);
      applyExports(env, site, call, result);
    } catch (err) {
      // eslint-disable-next-line @typescript-eslint/only-throw-error
      throw err instanceof Error ? err.message : String(err);
    }
  }
}

/**
 * Run every teardown call in order, each wrapped on its own: an error or an
 * unresolvable `$ref` skips that call and the rest still run.
 */
export async function runTeardown(env: ExecEnv, calls: Call[]): Promise<void> {
  for (const [i, call] of calls.entries()) {
    const site: FailureSite = {
      group: env.ctx.group,
      test: "<teardown>",
      scenarioFile: env.scenarioFile,
      step: `teardown[${i}]`,
    };
    try {
      const result = await performCall(env, site, call);
      applyExports(env, site, call, result);
    } catch (err) {
      env.log(
        `[${env.ctx.group}] teardown ${call.op} skipped: ${describeError(err)}`,
      );
    }
  }
}

// ─── A test ───────────────────────────────────────────────────────────────

/** Run one scenario test: the primary call, then every clause in order. */
export async function runScenarioTest(
  env: ExecEnv,
  test: ScenarioTest,
): Promise<void> {
  const site: FailureSite = {
    group: env.ctx.group,
    test: test.name,
    scenarioFile: env.scenarioFile,
    step: "call",
  };

  // A test with an errorCode clause *expects* the primary call to fail, and
  // the generator refuses such a test an export or a clause that reads the
  // primary response, so nothing downstream needs the response.
  const expectsError = test.assert.some((a) => a.kind === "errorCode");

  const primary: Primary = {
    op: test.call.op,
    params: {},
    response: undefined,
  };

  try {
    const result = await performCall(env, site, test.call);
    primary.params = result.params;
    primary.response = result.response;
    if (!expectsError) applyExports(env, site, test.call, result);
  } catch (err) {
    if (!expectsError || !isCallError(err)) throw err;
    const detail = callErrorDetail(err);
    primary.params = detail.params;
    primary.error = detail.cause;
  }

  for (const [i, clause] of test.assert.entries()) {
    await runAssertion(env, { ...site, step: `assert[${i}]` }, clause, primary);
  }
}

async function runAssertion(
  env: ExecEnv,
  site: FailureSite,
  clause: Assertion,
  primary: Primary,
): Promise<void> {
  switch (clause.kind) {
    case "responseField": {
      if (primary.error !== undefined) {
        throw fail(site, primary.op, primary.params, "responseField", {
          expected: "the call to succeed",
          actual: describeError(primary.error),
        });
      }
      const mismatch = guard(site, primary.op, primary.params, "responseField", () =>
        evaluateChecks(primary.response, clause.checks, env.ctx),
      );
      if (mismatch !== null) {
        throw fail(site, primary.op, primary.params, "responseField", mismatch);
      }
      return;
    }

    case "readback": {
      const result = await performCall(env, site, clause.call);
      const mismatch = guard(site, clause.call.op, result.params, "readback", () =>
        evaluateChecks(result.response, clause.checks, env.ctx),
      );
      if (mismatch !== null) {
        throw fail(site, clause.call.op, result.params, "readback", mismatch);
      }
      // "the call's exports are applied" — and only now, so a clause that
      // failed never overwrites a context value with a stale reading, and an
      // `eventually` applies them on the attempt that passes and only then.
      applyExports(env, site, clause.call, result);
      return;
    }

    case "listContains":
    case "absent": {
      // `absent` has two forms; only the error form carries `error`.
      if ("error" in clause) {
        await runAbsentError(env, site, clause.call, clause.error);
        return;
      }
      const source = clause.call
        ? await performCall(env, site, clause.call)
        : { params: primary.params, response: primary.response };
      const op = clause.call ? clause.call.op : primary.op;
      const mismatch = guard(site, op, source.params, clause.kind, () =>
        clause.kind === "listContains"
          ? evaluateListContains(source.response, clause.itemsPath, clause.where, env.ctx)
          : evaluateListAbsent(source.response, clause.itemsPath, clause.where, env.ctx),
      );
      if (mismatch !== null) {
        throw fail(site, op, source.params, clause.kind, mismatch);
      }
      if (clause.call) applyExports(env, site, clause.call, source);
      return;
    }

    case "errorCode": {
      if (primary.error === undefined) {
        throw fail(site, primary.op, primary.params, "errorCode", {
          expected: `an error (${acceptedCodes(clause.error)})`,
          actual: "<no error>",
        });
      }
      if (errorMatches(primary.error, clause.error)) return;
      throw fail(
        site,
        primary.op,
        primary.params,
        "errorCode",
        {
          expected: `an error (${acceptedCodes(clause.error)})`,
          actual: describeError(primary.error),
        },
        primary.error,
      );
    }

    case "eventually": {
      const delayMs = clause.delayMs ?? 0;
      const inner: FailureSite = { ...site, step: `${site.step}.assert` };
      let last: unknown;
      for (let attempt = 1; attempt <= clause.maxAttempts; attempt++) {
        try {
          await runAssertion(env, inner, clause.assert, primary);
          return;
        } catch (err) {
          last = err;
          if (attempt < clause.maxAttempts && delayMs > 0) await env.sleep(delayMs);
        }
      }
      // The last failure's message is the reported one; say how hard we tried.
      throw prefixAttempts(last, clause.maxAttempts, delayMs);
    }
  }
}

/** `absent`, error form: the call must fail with the named error. */
async function runAbsentError(
  env: ExecEnv,
  site: FailureSite,
  call: Call,
  spec: { shape: string; code: string },
): Promise<void> {
  let params: Record<string, unknown>;
  try {
    params = evaluateParams(call.params, env.ctx);
  } catch (err) {
    throw expressionFailure(site, call.op, err);
  }
  try {
    await env.send(call.op, params);
  } catch (err) {
    if (errorMatches(err, spec)) return;
    throw fail(
      site,
      call.op,
      params,
      "absent",
      {
        expected: `an error (${acceptedCodes(spec)})`,
        actual: describeError(err),
      },
      err,
    );
  }
  throw fail(site, call.op, params, "absent", {
    expected: `an error (${acceptedCodes(spec)})`,
    actual: "<no error>",
  });
}

// ─── Calls, exports and failures ──────────────────────────────────────────

const CALL_ERROR = Symbol.for("overcast.compat.scenario.callError");

/**
 * Evaluate a call's params and send it. An SDK error becomes a failure whose
 * message carries the six fields and whose cause is the original error, so
 * the harness's unimplemented detection still sees a 501.
 */
async function performCall(
  env: ExecEnv,
  site: FailureSite,
  call: Call,
): Promise<CallResult> {
  let params: Record<string, unknown>;
  try {
    params = evaluateParams(call.params, env.ctx);
  } catch (err) {
    throw expressionFailure(site, call.op, err);
  }
  try {
    // The response becomes the IR's document here, once, so a blob reads as
    // its base64 text in every clause and every export (toDocument).
    return { params, response: toDocument(await env.send(call.op, params)) };
  } catch (err) {
    const failure = fail(
      site,
      call.op,
      params,
      "call",
      { expected: "the call to succeed", actual: describeError(err) },
      err,
    );
    // Marked so runScenarioTest can tell "the primary call raised" (which an
    // errorCode clause expects) from every other kind of failure.
    Object.defineProperty(failure, CALL_ERROR, {
      value: { params, cause: err },
      enumerable: false,
    });
    throw failure;
  }
}

function isCallError(err: unknown): boolean {
  return err instanceof Error && CALL_ERROR in err;
}

interface CallErrorDetail {
  params: Record<string, unknown>;
  cause: unknown;
}

function callErrorDetail(err: unknown): CallErrorDetail {
  return (err as unknown as Record<symbol, CallErrorDetail>)[CALL_ERROR];
}

/**
 * Write a call's exports into the context bag. A path that does not resolve
 * is a failure for the step that carries it, naming the path.
 */
function applyExports(
  env: ExecEnv,
  site: FailureSite,
  call: Call,
  result: CallResult,
): void {
  if (!call.export) return;
  for (const [contextPath, responsePath] of Object.entries(call.export)) {
    const found = resolvePath(result.response, responsePath);
    if (!found.found) {
      throw fail(site, call.op, result.params, "export", {
        path: responsePath,
        expected: `the path to resolve, for the context value ${contextPath}`,
        actual: "<missing>",
      });
    }
    env.ctx.bag.set(contextPath, found.value);
  }
}

/**
 * Run a predicate that may throw an ExpressionError (an `equals` or a `where`
 * whose expected value is a `$ref`) and turn that into a step failure.
 */
function guard(
  site: FailureSite,
  op: string,
  params: Record<string, unknown>,
  kind: string,
  evaluate: () => Mismatch | null,
): Mismatch | null {
  try {
    return evaluate();
  } catch (err) {
    if (!(err instanceof ExpressionError)) throw err;
    throw fail(site, op, params, kind, {
      expected: "an evaluable expected value",
      actual: err.message,
    });
  }
}

function expressionFailure(
  site: FailureSite,
  op: string,
  err: unknown,
): Error {
  if (!(err instanceof ExpressionError)) return err instanceof Error ? err : new Error(String(err));
  return scenarioFailure(
    failureMessage(site, {
      op,
      params: "<not evaluated>",
      kind: "params",
      expected: "every parameter expression to evaluate",
      actual: err.message,
    }),
  );
}

function fail(
  site: FailureSite,
  op: string,
  params: unknown,
  kind: string,
  mismatch: Mismatch,
  cause?: unknown,
): Error {
  const detail = {
    op,
    params,
    kind,
    expected: mismatch.expected,
    actual: mismatch.actual,
    ...(mismatch.path === undefined ? {} : { path: mismatch.path }),
  };
  return scenarioFailure(failureMessage(site, detail), cause);
}

/**
 * Put the retry budget in front of the last attempt's message.
 *
 * The prefix is byte-for-byte the one compat/model/README.md § Failure
 * messages fixes, and the sibling interpreters emit the same string, so one
 * generated group's give-up reads identically whichever suite reports it and
 * a log search finds all three. It is a prefix rather than a suffix because
 * the six-field message ends in the scenario file and step index, which is
 * where a reader looks next.
 *
 * The message is rewritten in place rather than rethrown as a new Error: the
 * unimplemented markers scenarioFailure() copied onto this error are what
 * keeps a 501 inside an `eventually` classified as unimplemented, and a fresh
 * Error would drop them.
 */
function prefixAttempts(err: unknown, maxAttempts: number, delayMs: number): unknown {
  if (!(err instanceof Error)) return err;
  err.message =
    `eventually gave up after ${maxAttempts} attempt(s) ${delayMs}ms apart; ` +
    `last failure: ${err.message}`;
  return err;
}
